### PR TITLE
spanconfig: replace roachpb.SpanConfigEntry in package spanconfig 

### PR DESCRIPTION
**Pull request by arulajmani**
See individual commits for details. 

Polished up and carved out from https://github.com/cockroachdb/cockroach/pull/76138. 

In the next patch, we'll modify the existing span configuration RPC arguments to something similar in the draft PR above. Then, we'll make `spanconfig.Target` a union over system targets and spans (instead of the alias for roachpb.Spans as in this PR). We'll also add all the encoding and decoding methods for system targets that we've seen in prior draft patches.

### CODE REVIEW

File: pkg/spanconfig/spanconfigstore/spanconfigstore.go
Comment (irfansharif) [inline pkg/spanconfig/spanconfigstore/spanconfigstore.go:43]: I'm not going to hold up the merge over it, but I don't think this commit is a positive change. Thanks for separating it out into a stand-alone commit -- I think we should drop it. We have three variants now, and one in the future.

- spanconfig.Store;
- spanconfigstore.Store;
- spanconfigstore.spanConfigStore;
- (in the future) spanconfigstore.systemSpanConfigStore

This kind of spread, if anything, makes it difficult to talk about things meaninfully. Looking at the tests, I don't think the separation is also helping much. I understand the intent here was to make room for the stand-alone spanconfigstore.systemSpanConfigStore, but that too I don't think needs to be a wholly distinct type when instead it could be a set of fields on spanconfigstore.Store and appropriately named methods. The interface we care about for all the three types are spanconfig.StoreWriter (with it's diff-ing contract), primarily, and I think it's possible to test each component by interfacing through just the StoreWriter interface. Want to test spanconfigstore.systemSpanConfigStore? Just populate using the right kind of span.Targets. What to test spanconfigstore.systemSpanConfigStore? Do the same thing. Want to test the top-level spanconfigstore.Store? Again, just do the same thing.

Generally IMO when it's possible to test everything we care about through the high-level interface (which is the case here), I think we should. Peeking into the internals instead makes for test code that's going to be subject to more change over time than necessary.

**Review by irfansharif — APPROVED** [review]
These new types look great! I think framing the changes to this package in terms of spanconfig.{Target,Record} and in the future generalizing spanconfig.Target to also encompass the "special" targets is very clear IMO -- I dig it. I've left mostly minor comments below, with a recommendation that we abandon the third commit. I also have a question about the continued existence of roachpb.SpanConfigEntry, but nothing worth holding this PR on.

**Review by arulajmani — COMMENTED** [review]
TFTR! Will bors on green. 

edit -- missed a few comments above because they were over-written in a subsequent commit. Will merge on green and once those are resolved.

File: pkg/ccl/migrationccl/migrationsccl/seed_tenant_span_configs_external_test.go
Comment (arulajmani) [reply pkg/ccl/migrationccl/migrationsccl/seed_tenant_span_configs_external_test.go]: Done.

File: pkg/spanconfig/spanconfigstore/store.go
Comment (arulajmani) [reply pkg/spanconfig/spanconfigstore/store.go]: Yup, good call, done!

File: pkg/spanconfig/spanconfigstore/store.go
Comment (arulajmani) [reply pkg/spanconfig/spanconfigstore/store.go]: Done!

File: pkg/spanconfig/spanconfigstore/store_test.go
Comment (arulajmani) [reply pkg/spanconfig/spanconfigstore/store_test.go]: Done in the intermediate commit. Though, this does speak to my point of this being a test for the internal struct and not the interface/struct that implements the interface.

File: pkg/spanconfig/spanconfig.go
Comment (arulajmani) [reply pkg/spanconfig/spanconfig.go]: Changed.

**Comment by arulajmani** [comment]
It's green! TFTR, and thanks for brainstorming a bunch of this stuff to get to this point. I echo your sentiments around the new types! 

bors r+

**Pull request merged**
